document_page_reference/models/document_page.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2019 Creu Blanca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, tools, _
+from odoo.exceptions import ValidationError
+import logging
+_logger = logging.getLogger(__name__)
+
+try:
+    from jinja2.sandbox import SandboxedEnvironment
+    from jinja2 import Undefined
+
+    class Context(SandboxedEnvironment.context_class):
+        def resolve(self, key):
+            res = super().resolve(key)
+            if not isinstance(res, Undefined):
+                return res
+            return self.parent['ref'](key)
+
+    class Environment(SandboxedEnvironment):
+        context_class = Context
+
+    mako_template_env = Environment(
+        block_start_string="<%",
+        block_end_string="%>",
+        variable_start_string="${",
+        variable_end_string="}",
+        comment_start_string="<%doc>",
+        comment_end_string="</%doc>",
+        line_statement_prefix="%",
+        line_comment_prefix="##",
+        trim_blocks=True,               # do not output newline after blocks
+        autoescape=False,
+    )
+except Exception:
+    _logger.error("Jinja2 is not available")
+
+
+class DocumentPage(models.Model):
+
+    _inherit = 'document.page'
+
+    reference = fields.Char()
+    content_parsed = fields.Html(compute='_compute_content_parsed')
+
+    @api.depends('history_head')
+    def _compute_content_parsed(self):
+        for record in self:
+            record.content_parsed = record.get_content()
+
+    @api.constrains('reference')
+    def _check_reference(self):
+        for record in self:
+            if not record.reference:
+                continue
+            if self.search([
+                ('reference', '=', record.reference),
+                ('id', '!=', record.id)]
+            ):
+                raise ValidationError(_('Reference must be unique'))
+
+    def _get_document(self, code):
+        # Hook created in order to add check on other models
+        document = self.search([('reference', '=', code)])
+        if document:
+            return document
+        return False
+
+    def get_reference(self, code):
+        element = self._get_document(code)
+        if not element:
+            return code
+        if self.env.context.get('raw_reference', False):
+            return element.display_name
+        return '<a href="%s">%s</a>' % (
```

Comment:
Here you should probably use some regex or external libraries to remove html tags and prevent html injection

Reply:
Comments attended